File: challenges/strCount/c/strCount.c
/**
   challenges/strCount/c/strCount.c
   ====================================================
   CREATED: 2018-06-20
   UPDATED: 2018-07-05
   VERSION: v1.0.1
   AUTHOR: wlharvey4
   ABOUT: strCount challenge in C
   NOTES:
   CHANGE-LOG:
   ....................................................
   v1.0.1 2018-07-05T16:15:00
   convert return type to pointer
   ----------------------------------------------------
 */

# include "./strCount.h"

StrCount * strCount(Input input) {
  const char * str1 = input.str1;
  const char   str2 = input.str2;

  int * c = malloc(sizeof(StrCount));		/* result count */
  size_t offset = 0;	/* pointer into starting point of str1 for next search */
  char * p = NULL;      /* result pointer into str1 of search for str2 */

  while ((p = strchr(str1 + offset, str2))) { /* returns NULL on failure */
    offset = p - str1 + 1; 	/* reset the offset based upon found pointer value */
    (*c)++;
  }

  return c;
}

File: challenges/strCount/ruby/strCount.rb
#!/usr/bin/env ruby
# Code-Challenges-Intl/challenges/strCount/ruby/strCount.rb
# =============================================================================
# CREATED: 2018-10-23
# UPDATED: 2018-10-23
# VERSION: 1.1.0
# AUTHOR : wlharvey4
# ABOUT  : strCount challenge implemented in Ruby
# NOTES  : String#scan iterates through the string matching the pattern, adding
#	   the match to the result array
# CHANGE-LOG:
# .............................................................................
# version 1.0.0 2018-10-23T15:50
# -- initial implementation
# .............................................................................
# version 1.1.0 2018-10-23T16:20
# -- added #main method for command-line operation;
# -- refactored #inspect for better description
# -----------------------------------------------------------------------------

class StrCount
  attr_reader :str, :char, :count
  def initialize input
    @str   = input["str1"]
    @char  = input["str2"]
    @count = calculate @str, @char
  end
  def calculate str, char
    (str.scan /#{char}/).length
  end
  def to_s
    @count.to_s
  end
  def inspect
    "StrCount of '#{char}' in '#{str}' is #{count}"
  end
  def == c
    self.count == c
  end
end

def main str1, str2
  strCount = StrCount.new({"str1"=>str1,"str2"=>str2})
  p strCount
end

if $0 == __FILE__
  main $*[0], $*[1]
end
